File: 15-Algorithms_dynamic_programming/fibonacci.js
const fibonacciMaster = () => {
  let cache = {};

  return function fib(n) {
    if (n in cache) return cache[n];
    if (n < 2) return n;

    cache[n] = fib(n - 1) + fib(n - 2);
    return cache[n];
  };
};

const fasterFib = fibonacciMaster();
console.log(fasterFib(10));

File: 15-Algorithms_dynamic_programming/fibonacci.py
def fibonacci():
    cache = {}

    def fib(n):
        if n in cache:
            return cache[n]

        if n < 2:
            return n

        cache[n] = fib(n - 1) + fib(n - 2)
        return cache[n]

    return fib


faster_fib = fibonacci()
print(faster_fib(10))
